### PR TITLE
Feature/doc 178

### PR DESCRIPTION
**Pull request by electronicpeter**
Main features of old docusafe business-layer have been implemented:
- create/destroy users
- create/delete/update documents
- delete whole folders
- check user exists
- check document exists

Not yet implemented are 
- inbox management
- dfs registration

And the whole transactional layer including version numbers.

Which of those features will be uses must be cleared with the datev team first. This adapter is working stable and thus should be merged to develop soon. Further features will be implemented with a new issue.

Code coverage is now slightly above 90%. If this is still the case after merge, the rate will be the new lower limit.

### CODE REVIEW

File: datasafe-simple-adapter/datasafe-simple-adapter-impl/src/main/java/de/adorsys/datasafe/simple/adapter/impl/SwitchablePathEncryptionImpl.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    private static boolean withPathEncryption = checkIsPathEncryptionToUse();
+
+    @Inject
+    public SwitchablePathEncryptionImpl(SymmetricPathEncryptionService bucketPathEncryptionService, PrivateKeyService privateKeyService) {
+        super(bucketPathEncryptionService, privateKeyService);
+    }
+
+    public static boolean checkIsPathEncryptionToUse() {
+        if (System.getProperty(NO_BUCKETPATH_ENCRYPTION) != null) {
```
Comment (valb3r) [inline datasafe-simple-adapter/datasafe-simple-adapter-impl/src/main/java/de/adorsys/datasafe/simple/adapter/impl/SwitchablePathEncryptionImpl.java]: maybe extracting local variable before if 
`String value = System.getProperty(NO_BUCKETPATH_ENCRYPTION);`
could be better - no double access to system-properties

File: datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        Assertions.assertEquals(15, listFound.size());
+        DocumentFQN otherDoc = new DocumentFQN("affe/subdir_0/subdir_0/file0txt");
+        simpleDatasafeService.deleteDocument(userIDAuth, otherDoc);
+        listFound = simpleDatasafeService.list(userIDAuth, root, ListRecursiveFlag.TRUE);
+        show("full list recursive after delete one file", listFound);
+        Assertions.assertEquals(14, listFound.size());
+    }
+
+    @ParameterizedTest
+    @MethodSource({"parameterCombination"})
```
Comment (valb3r) [inline datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java]: Curly braces not necessary

File: datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+        // now read the docs with the correct password
+        DSDocument dsDocument = simpleDatasafeService.readDocument(userIDAuth, new DocumentFQN(path));
+        Assertions.assertArrayEquals(content.getBytes(), dsDocument.getDocumentContent().getValue());
+
+        DSDocument dsDocument2 = simpleDatasafeService.readDocument(userIDAuth2, new DocumentFQN(path));
+        Assertions.assertArrayEquals(content.getBytes(), dsDocument2.getDocumentContent().getValue());
+
+        simpleDatasafeService.destroyUser(userIDAuth2);
+        Assertions.assertFalse(simpleDatasafeService.documentExists(userIDAuth2, document.getDocumentFQN()));
```
Comment (valb3r) [inline datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java]: maybe it makes sense to import static for `Assertions`

File: datasafe-types-api/src/test/java/de/adorsys/datasafe/types/api/de/adorsys/datasafe/types/api/resource/UriTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.net.URI;
+
+public class UriTest {
+    @Test
+    public void anUriResolveTest() {
+        String uriString = "http://Uri.peter.at.home/";
+        Uri uri = new Uri(uriString);
+        uri = uri.resolve("a/").resolve("b/");
+        Assertions.assertEquals(uriString + "a/b/", uri.toASCIIString());
```
Comment (valb3r) [inline datasafe-types-api/src/test/java/de/adorsys/datasafe/types/api/de/adorsys/datasafe/types/api/resource/UriTest.java]: static import `Assertions` will improve readability

File: datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    @ValueSource
+    protected static Stream<Boolean> withOrWithoutEncryption() {
+        return Stream.of(
+                Boolean.TRUE,
+                Boolean.FALSE);
+    }
+
+    private static Stream<Arguments> parameterCombination() {
```
Comment (valb3r) [inline datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java]: Can be done slightly shorter:
```
private static Stream<Arguments> parameterCombination() {
        return Sets.cartesianProduct(
                allStorages().collect(Collectors.toSet()),
                ImmutableSet.of(true, false)
        ).stream().map(it -> Arguments.of(it.get(0), it.get(1)));
    }
```

File: datasafe-business/src/test/java/de/adorsys/datasafe/business/impl/e2e/WithStorageProvider.java [hunk 6]
```diff
@@ -258,16 +305,32 @@ private static void startCeph() {
...
-        StorageDescriptor(String name, Supplier<StorageService> storageService, Uri location) {
-            this.name = name;
-            this.storageService = storageService;
-            this.location = location;
+        private final String accessKey;
+        private final String secretKey;
+        private final String region;
+        private final String rootBucket;
+
+        public String getMappedUrl() {
```
[the remark this answers is not in the document]
Comment (electronicpeter) [reply datasafe-business/src/test/java/de/adorsys/datasafe/business/impl/e2e/WithStorageProvider.java]: Maybe we discuss this. I did not understand why Maksym made all methods static. The mapped URL is not known unless the storage is wrapped into suppliers which are not instantiated unless they are used. So, yes, it is not thread safe. But for my tests it worked fine.

File: datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java (hunk 5 above)
Comment (electronicpeter) [reply datasafe-simple-adapter/datasafe-simple-adapter-impl/src/test/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeAdapterTest.java]: cool!

**Pull request merged**